implement releaserequest for non-maintenance prj

lethliel opened the pull request:
In the API a new request action release was implemented. This changes
enables the user to create a release request for non-maintenance projects
and to review / view the release requests.

jira#OBS-51
https://github.com/openSUSE/open-build-service/pull/9943/files

@marcus-h: please review

marcus-h commented:
LGTM (almost)!:)

On 2020-08-05 12:23:50 -0700, Marco Strigl wrote:
> Subject: [PATCH] implement releaserequest for non-maintenance prj
>
> In the API a new request action release was implemented. This changes
> enable the user to create a release request for non-maintenance projects

Minor: enables

> and to review / view the release requests
> ---
>  osc/commandline.py | 35 +++++++++++++++++++++++++++++------
>  osc/core.py        |  7 +++++++
>  2 files changed, 36 insertions(+), 6 deletions(-)
>
> diff --git a/osc/commandline.py b/osc/commandline.py
> index 68d4db5e..3c901a83 100644
> --- a/osc/commandline.py
> +++ b/osc/commandline.py
> @@ -3107,18 +3107,28 @@ def do_release(self, subcmd, opts, *args):
>
>
>      @cmdln.option('-m', '--message', metavar='TEXT',
> -                        help='specify message TEXT')
> +                  help='specify message TEXT')
> +    @cmdln.option('-p', '--package', metavar='DIR', action='append',

Hmm 'DIR' sounds a bit strange. What about metavar='PAC' or metavar='PKG'?

> +                  help='specify packages to release')
>      def do_releaserequest(self, subcmd, opts, *args):
> -        """${cmd_name}: Create a request for releasing a maintenance update.
> +        """${cmd_name}: Create a release request
>
> -        [See http://openbuildservice.org/help/manuals/obs-reference-guide/cha.obs.maintenance_setup.html
> -         for information on this topic.]
> +        For maintenance incident projects:
>
>          This command is used by the maintenance team to start the release process of a maintenance update.
>          This includes usually testing based on the defined reviewers of the update project.
>
> +        [See http://openbuildservice.org/help/manuals/obs-reference-guide/cha.obs.maintenance_setup.html
> +         for information on this topic.]
> +
> +        For normal projects:
> +
> +        This command is used to transfer sources and binaries without rebuilding them.
> +        It requires defined release targets set to trigger="manual". Please refer the
> +        release management chapter in the OBS book for details.
> +

Hmm what about a link?:) (If the chapter does not exist yet, let's remove
the sentence.)

>          usage:
> -            osc releaserequest [ SOURCEPROJECT ]
> +            osc releaserequest [-p package] [ SOURCEPROJECT ]
>
>          ${cmd_option_list}
>          """
> @@ -3140,12 +3150,25 @@ def do_releaserequest(self, subcmd, opts, *args):
>          if len(args) > 0:
>              source_project = args[0]
>
> +        f = show_project_meta(apiurl, source_project)
> +        root = ET.fromstring(b''.join(f))
> +
>          if not opts.message:
>              opts.message = edit_message()
>          else:
>              opts.message = str(opts.message.encode().decode('unicode_escape'))
>
> -        r = create_release_request(apiurl, source_project, opts.message)
> +        if 'kind' in root.attrib and root.attrib['kind'] == 'maintenance_incident':
> +            r = create_release_request(apiurl, source_project, opts.message)
> +        else:
> +            r = Request()
> +            if opts.package:
> +                for pac in opts.package:
> +                    r.add_action('release', src_project=source_project, src_package=pac)
> +            else:
> +                r.add_action('release', src_project=source_project)
> +            r.description = _html_escape(unicode(opts.message, 'utf8'))

Hmm why the unicode(...)?


> diff --git a/osc/core.py b/osc/core.py
> index 6613eed0..895a24b8 100644
> --- a/osc/core.py
> +++ b/osc/core.py
> @@ -2941,6 +2944,10 @@ def prj_pkg_join(prj, pkg, repository=None):
>                  d['source'] = d['source'] + '@%s' % action.src_rev
>              d['source'] = d['source'] + ' ->'
>              d['target'] = prj_pkg_join(action.tgt_project, action.tgt_package)
> +        elif action.type == 'release':
> +            d['source'] = '%s' % prj_pkg_join(action.src_project, action.src_package)

There's not need for str interpolation:)

> +            d['source'] = d['source'] + ' ->'
> +            d['target'] = prj_pkg_join(action.tgt_project, action.tgt_package)

adrianschroeter commented:
The book link for maintenance has changed to 

  https://openbuildservice.org/help/manuals/obs-user-guide/cha.obs.maintenance_setup.html

The release request will appear here

 https://openbuildservice.org/help/manuals/obs-user-guide/cha.obs.request_and_review_ystem.html

lethliel commented:
On 8/5/20 11:55 PM, Marcus Hüwe wrote:
> LGTM (almost)!:)
> 
> On 2020-08-05 12:23:50 -0700, Marco Strigl wrote:
>> Subject: [PATCH] implement releaserequest for non-maintenance prj
>>
>> In the API a new request action release was implemented. This changes
>> enable the user to create a release request for non-maintenance projects
> 
> Minor: enables
> 

fixed

>> and to review / view the release requests
>> ---
>> osc/commandline.py | 35 +++++++++++++++++++++++++++++------
>> osc/core.py | 7 +++++++
>> 2 files changed, 36 insertions(+), 6 deletions(-)
>>
>> diff --git a/osc/commandline.py b/osc/commandline.py
>> index 68d4db5e..3c901a83 100644
>> --- a/osc/commandline.py
>> +++ b/osc/commandline.py
>> @@ -3107,18 +3107,28 @@ def do_release(self, subcmd, opts, *args):
>>
>>
>> @cmdln.option('-m', '--message', metavar='TEXT',
>> - help='specify message TEXT')
>> + help='specify message TEXT')
>> + @cmdln.option('-p', '--package', metavar='DIR', action='append',
> 
> Hmm 'DIR' sounds a bit strange. What about metavar='PAC' or metavar='PKG'?
> 

Yes. You are right.

>> + help='specify packages to release')
>> def do_releaserequest(self, subcmd, opts, *args):
>> - """${cmd_name}: Create a request for releasing a maintenance update.
>> + """${cmd_name}: Create a release request
>>
>> - [See
> http://openbuildservice.org/help/manuals/obs-reference-guide/cha.obs.maintenance_setup.html
>> - for information on this topic.]
>> + For maintenance incident projects:
>>
>> This command is used by the maintenance team to start the release
> process of a maintenance update.
>> This includes usually testing based on the defined reviewers of the
> update project.
>>
>> + [See
> http://openbuildservice.org/help/manuals/obs-reference-guide/cha.obs.maintenance_setup.html
>> + for information on this topic.]
>> +
>> + For normal projects:
>> +
>> + This command is used to transfer sources and binaries without
> rebuilding them.
>> + It requires defined release targets set to trigger="manual". Please
> refer the
>> + release management chapter in the OBS book for details.
>> +
> 
> Hmm what about a link?:) (If the chapter does not exist yet, let's remove
> the sentence.)
> 

added the link and corrected the link to the maintenance setup.

>> usage:
>> - osc releaserequest [ SOURCEPROJECT ]
>> + osc releaserequest [-p package] [ SOURCEPROJECT ]
>>
>> ${cmd_option_list}
>> """
>> @@ -3140,12 +3150,25 @@ def do_releaserequest(self, subcmd, opts, *args):
>> if len(args) > 0:
>> source_project = args[0]
>>
>> + f = show_project_meta(apiurl, source_project)
>> + root = ET.fromstring(b''.join(f))
>> +
>> if not opts.message:
>> opts.message = edit_message()
>> else:
>> opts.message = str(opts.message.encode().decode('unicode_escape'))
>>
>> - r = create_release_request(apiurl, source_project, opts.message)
>> + if 'kind' in root.attrib and root.attrib['kind'] ==
> 'maintenance_incident':
>> + r = create_release_request(apiurl, source_project, opts.message)
>> + else:
>> + r = Request()
>> + if opts.package:
>> + for pac in opts.package:
>> + r.add_action('release', src_project=source_project, src_package=pac)
>> + else:
>> + r.add_action('release', src_project=source_project)
>> + r.description = _html_escape(unicode(opts.message, 'utf8'))
> 
> Hmm why the unicode(...)?
> 

I just copied the existing line. But you are right. This is old python2
stuff and is not needed anymore. Removed.

> 
>> diff --git a/osc/core.py b/osc/core.py
>> index 6613eed0..895a24b8 100644
>> --- a/osc/core.py
>> +++ b/osc/core.py
>> @@ -2941,6 +2944,10 @@ def prj_pkg_join(prj, pkg, repository=None):
>> d['source'] = d['source'] + '@%s' % action.src_rev
>> d['source'] = d['source'] + ' ->'
>> d['target'] = prj_pkg_join(action.tgt_project, action.tgt_package)
>> + elif action.type == 'release':
>> + d['source'] = '%s' % prj_pkg_join(action.src_project,
> action.src_package)
> 
> There's not need for str interpolation:)
>

Force of habit ;)

Thanks for the review.

marcus-h commented:
On 2020-08-06 02:00:56 -0700, Marco Strigl wrote:
> On 8/5/20 11:55 PM, Marcus Hüwe wrote:

<SNIP>

> >> diff --git a/osc/core.py b/osc/core.py
> >> index 6613eed0..895a24b8 100644
> >> --- a/osc/core.py
> >> +++ b/osc/core.py
> >> @@ -2941,6 +2944,10 @@ def prj_pkg_join(prj, pkg, repository=None):
> >> d['source'] = d['source'] + '@%s' % action.src_rev
> >> d['source'] = d['source'] + ' ->'
> >> d['target'] = prj_pkg_join(action.tgt_project, action.tgt_package)
> >> + elif action.type == 'release':
> >> + d['source'] = '%s' % prj_pkg_join(action.src_project,
> > action.src_package)
> >
> > There's not need for str interpolation:)
> >
>
> Force of habit ;)
>
It's still there:P Now the diff looks like this:

> @@ -2936,11 +2939,15 @@ def prj_pkg_join(prj, pkg, repository=None):
>              if action.opt_sourceupdate and show_srcupdate:
>                  srcupdate = '(%s)' % action.opt_sourceupdate
>          elif action.type == 'maintenance_release':
> -            d['source'] = '%s' % prj_pkg_join(action.src_project, action.src_package)
> +            d['source'] = prj_pkg_join(action.src_project, action.src_package)
>              if action.src_rev:
>                  d['source'] = d['source'] + '@%s' % action.src_rev
>              d['source'] = d['source'] + ' ->'
>              d['target'] = prj_pkg_join(action.tgt_project, action.tgt_package)
> +        elif action.type == 'release':
> +            d['source'] = '%s' % prj_pkg_join(action.src_project, action.src_package)
> +            d['source'] = d['source'] + ' ->'
> +            d['target'] = prj_pkg_join(action.tgt_project, action.tgt_package)

Btw., should we consider the action.src_rev in the "release" case, too?

adrianschroeter commented:
On Donnerstag, 6. August 2020, 13:16:34 CEST wrote Marcus Hüwe:
> On 2020-08-06 02:00:56 -0700, Marco Strigl wrote:
> > On 8/5/20 11:55 PM, Marcus Hüwe wrote:
> 
> <SNIP>
..
> > @@ -2936,11 +2939,15 @@ def prj_pkg_join(prj, pkg, repository=None):
> >              if action.opt_sourceupdate and show_srcupdate:
> >                  srcupdate = '(%s)' % action.opt_sourceupdate
> >          elif action.type == 'maintenance_release':
> > -            d['source'] = '%s' % prj_pkg_join(action.src_project, action.src_package)
> > +            d['source'] = prj_pkg_join(action.src_project, action.src_package)
> >              if action.src_rev:
> >                  d['source'] = d['source'] + '@%s' % action.src_rev
> >              d['source'] = d['source'] + ' ->'
> >              d['target'] = prj_pkg_join(action.tgt_project, action.tgt_package)
> > +        elif action.type == 'release':
> > +            d['source'] = '%s' % prj_pkg_join(action.src_project, action.src_package)
> > +            d['source'] = d['source'] + ' ->'
> > +            d['target'] = prj_pkg_join(action.tgt_project, action.tgt_package)
> 
> Btw., should we consider the action.src_rev in the "release" case, too?

it should not be set initially, but can be set by the server on creation.

(because all revisions except the current ones would not match the binaries)

-- 

Adrian Schroeter
email: adrian@suse.de

SUSE Linux GmbH, GF: Felix Imendörffer, Jane Smithard, Graham Norton, HRB 21284 (AG Nürnberg) 
 
Maxfeldstraße 5                         
90409 Nürnberg 
Germany

lethliel commented:
@marcus-h 

Oops. Was in the wrong if case.

marcus-h commented:
On 2020-08-06 04:25:02 -0700, Adrian Schröter wrote:
> On Donnerstag, 6. August 2020, 13:16:34 CEST wrote Marcus Hüwe:
> > On 2020-08-06 02:00:56 -0700, Marco Strigl wrote:
> > > On 8/5/20 11:55 PM, Marcus Hüwe wrote:
> >
> > <SNIP>
> ..
> > > @@ -2936,11 +2939,15 @@ def prj_pkg_join(prj, pkg, repository=None):
> > >              if action.opt_sourceupdate and show_srcupdate:
> > >                  srcupdate = '(%s)' % action.opt_sourceupdate
> > >          elif action.type == 'maintenance_release':
> > > -            d['source'] = '%s' % prj_pkg_join(action.src_project, action.src_package)
> > > +            d['source'] = prj_pkg_join(action.src_project, action.src_package)
> > >              if action.src_rev:
> > >                  d['source'] = d['source'] + '@%s' % action.src_rev
> > >              d['source'] = d['source'] + ' ->'
> > >              d['target'] = prj_pkg_join(action.tgt_project, action.tgt_package)
> > > +        elif action.type == 'release':
> > > +            d['source'] = '%s' % prj_pkg_join(action.src_project, action.src_package)
> > > +            d['source'] = d['source'] + ' ->'
> > > +            d['target'] = prj_pkg_join(action.tgt_project, action.tgt_package)
> >
> > Btw., should we consider the action.src_rev in the "release" case, too?
>
> it should not be set initially, but can be set by the server on creation.
>
That is, it makes sense to display it (if it is set)? If so, we can simply
unify the new codepath with the "maintenance_release" case.

if action.type in ('maintenance_release', 'release'):
    ...

> (because all revisions except the current ones would not match the binaries)
>
Ok. Maybe a stupid question: why isn't this simply part of the acceptinfo
then?

adrianschroeter commented:
On Donnerstag, 6. August 2020, 15:41:28 CEST wrote Marcus Hüwe:
> On 2020-08-06 04:25:02 -0700, Adrian Schröter wrote:
> > On Donnerstag, 6. August 2020, 13:16:34 CEST wrote Marcus Hüwe:
> > > On 2020-08-06 02:00:56 -0700, Marco Strigl wrote:
> > > > On 8/5/20 11:55 PM, Marcus Hüwe wrote:
> > >
> > > <SNIP>
> > ..
> > > > @@ -2936,11 +2939,15 @@ def prj_pkg_join(prj, pkg, repository=None):
> > > >              if action.opt_sourceupdate and show_srcupdate:
> > > >                  srcupdate = '(%s)' % action.opt_sourceupdate
> > > >          elif action.type == 'maintenance_release':
> > > > -            d['source'] = '%s' % prj_pkg_join(action.src_project, action.src_package)
> > > > +            d['source'] = prj_pkg_join(action.src_project, action.src_package)
> > > >              if action.src_rev:
> > > >                  d['source'] = d['source'] + '@%s' % action.src_rev
> > > >              d['source'] = d['source'] + ' ->'
> > > >              d['target'] = prj_pkg_join(action.tgt_project, action.tgt_package)
> > > > +        elif action.type == 'release':
> > > > +            d['source'] = '%s' % prj_pkg_join(action.src_project, action.src_package)
> > > > +            d['source'] = d['source'] + ' ->'
> > > > +            d['target'] = prj_pkg_join(action.tgt_project, action.tgt_package)
> > >
> > > Btw., should we consider the action.src_rev in the "release" case, too?
> >
> > it should not be set initially, but can be set by the server on creation.
> >
> That is, it makes sense to display it (if it is set)? If so, we can simply
> unify the new codepath with the "maintenance_release" case.

yes, release and maintenance_release are the same here.

> if action.type in ('maintenance_release', 'release'):
>     ...
> 
> > (because all revisions except the current ones would not match the binaries)
> >
> Ok. Maybe a stupid question: why isn't this simply part of the acceptinfo
> then?

acceptinfo gets only created after accept. I would not change this and add an
acceptinfo already on request creation...

-- 

Adrian Schroeter
email: adrian@suse.de

SUSE Linux GmbH, GF: Felix Imendörffer, Jane Smithard, Graham Norton, HRB 21284 (AG Nürnberg) 
 
Maxfeldstraße 5                         
90409 Nürnberg 
Germany

lethliel commented:
This is needed to test https://github.com/openSUSE/open-build-service/pull/9943 internally. 
So this will be merged before the API PR is merged.
